eval/adjust.rkt

```diff
@@ -71,333 +71,263 @@
     (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers)) ; tail of the current instr
     (define output (vector-ref vregs n))                            ; output of the current instr
     
-    (define exps-from-above (vector-ref vprecs-new (- n varc)))     ; vprecs-new is shifted by varc elements from vregs
-    (define new-exponents (get-exponent op output srcs))
+    (define intro (vector-ref vprecs-new (- n varc)))               ; intro for the current instruction
+    (define ampls (get-ampls op output srcs))                       ; ampls for the tail instructions
 
-    (define final-parent-precision (max (+ exps-from-above
+    (define final-parent-precision (max (+ intro
                                            (vector-ref vstart-precs (- n varc)))
                                         (*base-tuning-precision*)))
-
-    ; This case is weird. if we have a cancellation in fma -> ival-mult in fma should be in higher precision
-    (when (equal? op ival-fma)
-      (set! final-parent-precision (+ final-parent-precision (car new-exponents))))
     
-    (when (>= final-parent-precision (*rival-max-precision*))         ; Early stopping
+    (when (>= final-parent-precision (*rival-max-precision*))       ; Early stopping
       (*sampling-iteration* (*rival-max-iterations*)))
+    
+    ; Final precision assignment
     (vector-set! vprecs-new (- n varc) (min final-parent-precision (*rival-max-precision*)))
 
+    ; Intro and ampl propogation for each tail instruction
     (for ([x (in-list tail-registers)]
-          [new-exp (in-list new-exponents)]
+          [ampl (in-list ampls)]
           #:when (>= x varc)) ; when tail register is not a variable
       ; check whether this op already has a precision that is higher
-      (when (> (+ exps-from-above new-exp) (vector-ref vprecs-new (- x varc)))
-        (vector-set! vprecs-new (- x varc) (+ exps-from-above new-exp))))))
+      (when (> (+ intro ampl) (vector-ref vprecs-new (- x varc)))
+        (vector-set! vprecs-new (- x varc) (+ intro ampl))))))
+
+(define (exp-is-inf? x)
+  (equal? x 9223372036854775805))
+(define (exp-is-0? x)
+  (equal? x -9223372036854775807))
+; this function imitates an exponent of infinity as 9223372036854775805
+(define (replace-inf x)
+  (if (equal? x -9223372036854775805)
+      (abs x)
+      x))
```

Comment:
I'm confused why / when this comes up. This seems like it's just going to misidentify some very small numbers as being infinite?

Reply:
Oh, this is my weird thing. Exponent of `inf.bf` is equal to `-9223372036854775805` and I wanted to change it to a positive value instead to mimic actual infinity. This is useful when I do `maxlog`, `minlog` stuff. I do not use it anywhere else. Idk, this absolute value is just handy for these 2 functions